### PR TITLE
Add  min (Z_MIN) and max (Z_MAX) macros with single evaluation of arguments

### PR DESCRIPTION
**Pull request by nordic-krch**
As #17844 revealed usage of `MAX` and `MIN` macros results in bigger code and potential bug because arguments are evaluated twice. PR proposes alternative macros (`Z_MAX` and `Z_MIN`). Due to usage limitations they are not 100% replacements (thus new macros and not new implementation).

`Z_MAX` and `Z_MIN` cannot be used in cases like:
- constant integer generation (e.g. `__aligned(Z_MAX(4,5))`
- static variables (e.g. array for stack `static u8_t array[Z_MAX(...)];`)

### CODE REVIEW

File: include/sys/util.h
Comment (ioannisg) [inline include/sys/util.h]: larger

**Review by ioannisg — COMMENTED** [review]
Thanks, looks fine - I left a doc comment

File: include/sys/util.h
Comment (nordic-krch) [reply include/sys/util.h]: done

**Review by ioannisg — APPROVED** [review]

File: include/sys/util.h
Comment (andyross) [inline include/sys/util.h]: Random!?  It's the same every time!

I'll wait for @ceolin to chime in on the need for MISRA tool analysis of the RNG implementation that produced this.

No seriously, this is clever and whimsical and I love it.

**Review by andyross — APPROVED** [review]
This is fine with me, but a reasonably serious complaint could be made that gcc extensions like statement expressions shouldn't be in include/sys/util.h but instead in a toolchain-specific header, with a fallback for traditional preprocessors (and, heh, test coverage for that somehow even on gcc systems...).

Maybe not worth it given the simplicity of the patch, but something to think about.

File: include/sys/util.h
Comment (jukkar) [inline include/sys/util.h]: Your PR description says
```
Z_MAX and Z_MIN cannot be used in cases like:

    constant integer generation (e.g. __aligned(Z_MAX(4,5))
    static variables (e.g. array for stack static u8_t array[Z_MAX(...)];)
```
Why not add these comments also here to make it more clear when the macros cannot be used.

File: include/sys/util.h
Comment (andrewboie) [reply include/sys/util.h]: aren't these declarations totally within the scope of this block anyway? why do we need this?

**Review by tyrantlizard77 — DISMISSED** [review]
please get rid of these crazy variable names, they aren't needed

File: include/sys/util.h
Comment (pabigot) [reply include/sys/util.h]: @andrewboie The concern is that either macro argument `a` or `b` may expand to text that was intended to reference a variable with the chosen name in the enclosing scope.

A reasonably common alternative is to just [use a variable that ends with an underscore](https://gcc.gnu.org/onlinedocs/gcc-9.1.0/cpp/Duplication-of-Side-Effects.html#Duplication-of-Side-Effects).  `value_a_` would be a more maintainable choice as it more clearly identifies what it holds and is unlikely to conflict.

**Review by pabigot — DISMISSED** [review]
Too bad C++ `constexpr` isn't available, as it solves this problem very nicely.

This comment isn't part of the blocker, but I also dislike GNU extensions in supposedly portable code, and would rather developers and reviewers understand that macro arguments might be evaluated multiple times so high-cost calculations should be captured externally.  Is there no policy on this in Zephyr? (edit: "this" referring to use of GNU extensions.)

The blocker from me is, like @andrewboie, the variable names that bear no relation to the parameters they capture.  See comment for alternative.

**Comment by andrewboie** [comment]
> This is fine with me, but a reasonably serious complaint could be made that gcc extensions like statement expressions shouldn't be in include/sys/util.h but instead in a toolchain-specific header, with a fallback for traditional preprocessors (and, heh, test coverage for that somehow even on gcc systems...).

Good point.
Can you put in include/toolchain/gcc.h instead (which gets auto-included by toolchain.h)

**Comment by nordic-krch** [comment]
@andrewboie 
>please get rid of these crazy variable names, they aren't needed

as @pabigot mentioned it is used to ensure that there will be no variable name conflict. If they are named `_a` and `_b` and somebody calls `Z_MAX(4,_a)` macro will return wrong value. @pabigot I went for random names because there is always a risk that chosen prefix will be create variable name that already exists in given scope (e,g, `Z_MAX(a, a_)`)

**Comment by nordic-krch** [comment]
Macros moved to `gcc.h`. In `util.h` added comment for `MAX` and `MIN` about double evaluation and reference to single evaluation macros.

**Comment by nordic-krch** [comment]
@dbkinder can you help me with documentation error i'm getting:
`/var/lib/shippable/build/IN/main_repo/zephyr/include/sys/util.h:92: warning: unable to resolve reference to `Z_MAX' for \ref command`

I have `Z_MAX` defined in `include/toolchain/gcc.h`. Is that one not processed by doxygen?

**Comment by pabigot** [comment]
> It's also not an option to create variable based on input (`a ## _value`) because `a` might be an expression.

@nordic-krch You misunderstood my request.  I want to see:
```
#define Z_MIN(a, b) ({ \
￼			__typeof__ (a) value_a_ = (a); \
￼			__typeof__ (b) value_b_ = (b); \
￼			value_a_ < value_b_ ? value_a_ : value_b_; \
￼		})
```

If you are adamant about reducing readability against an infinitesimal chance of conflict, the minimum I would acccept is:

```
#define Z_MIN(a, b) ({ \
￼			/* random variable names to avoid naming conflict */ \
￼			__typeof__ (a) a_jjucpign = (a); \
￼			__typeof__ (b) b_jjucpign = (b); \
￼			a_jjucpign <  b_jjucpign ? a_jjucpign : b_jjucpign; \
￼		})
```

which at least lets a maintainer trace the in-macro value back to a parameter without having the cognitive load of dealing with totally meaningless identifiers.

**Comment by nordic-krch** [comment]
@pabigot applied your suggestion. Documentation issue pending (@dbkinder can you help?)

File: include/toolchain/gcc.h
Comment (pabigot) [inline include/toolchain/gcc.h]: `b_jjucpign`

Comment (nordic-krch) [reply include/toolchain/gcc.h]: 😳 fixed.

**Comment by nordic-krch** [comment]
> I would really prefer the crazy names replaced

done

> include/toolchain/gcc.h is that one not processed by doxygen?

Documentation issue is blocking this PR, @dbkinder can you help?

File: include/toolchain/gcc.h
Comment (pabigot) [inline include/toolchain/gcc.h]: While C allows double-underscore to appear anywhere except the start of an identifier, in C++ a double underscore anywhere in the identifier invoked undefined behavior.  This header will be included in C++ compilations.  Please remove the second underscore at the end of these identifiers.

**Review by pabigot — CHANGES_REQUESTED** [review]
Identifier names need to be modified for C++ compatibility.

**Comment by andrewboie** [comment]
Other than @pabigot comment on C++ naming this LGTM

File: include/toolchain/gcc.h
Comment (nordic-krch) [reply include/toolchain/gcc.h]: done

Comment (nordic-krch) [reply include/toolchain/gcc.h]: ping

**Review by pabigot — APPROVED** [review]

**Comment by nordic-krch** [comment]
@nashif @andrewboie @ioannisg ready for merge alert.

**Pull request merged**
